### PR TITLE
GAMBLR Improvements and Bugfixes

### PR DESCRIPTION
**Pull request by mattssca**
This PR is providing a fix for many of the active issues on GAMBLR, as well as other quality-of-life improvements. The following issues have been addressed in this PR:

- https://github.com/morinlab/GAMBLR/issues/161
- https://github.com/morinlab/GAMBLR/issues/160
- https://github.com/morinlab/GAMBLR/issues/159
- https://github.com/morinlab/GAMBLR/issues/157
- https://github.com/morinlab/GAMBLR/issues/156
- https://github.com/morinlab/GAMBLR/issues/153
- https://github.com/morinlab/GAMBLR/issues/128
- https://github.com/morinlab/GAMBLR/issues/102

- This PR also includes updates that are ensuring that the thematic vignettes can be knitted in remote sessions as well.
- Many updates in this PR are related to the function documentation overhaul.
- All functions that are using the config in any way to get file paths have been updated to encapsulate these statements in the newly added helper function that checks the config keys.
- README.md has been updated with a new section called "For Developers". This section details what the best practices are for designing and adding a new function to GAMBLR, complete with examples and a function template that can be used for designing new functions as well as an example of an adequately documented GAMBLR function.
- Minor hot fixes prevented the vignettes from being knitted. For some reason, the vignettes errors out when a warning gets thrown. Even if the `warnings` in the chunk option is set to FALSE.
- This branch was tested by running the vignettes as well as the [test_funcitons.R](https://github.com/morinlab/GAMBLR/blob/master/test_functions.R) script.
- the vignettes have also been updated to rely on modified code.
- The package documentation has also been re-generated to match all the updates included in this PR.

# Pull Request Checklists

**Important:** When opening a pull request, keep only the applicable checklist and delete all other sections.

### Required

- [x] I tested the new code for my use case (please provide a reproducible example of how you tested the new functionality)

- [x] I ensured all dplyr functions that commonly conflict with other packages are fully qualified. 

This can be checked and addressed by running `check_functions.pl` and responding to the prompts. Test your code _after_ you do this.

- [x] I generated the documentation and checked for errors relating to the new function (e.g. `devtools::document()`) and added `NAMESPACE` and all other modified files in the root directory and under `man`. 

### Optional but preferred with PRs

- [x] I updated and/or successfully knitted a vignette that relies on the modified code (which ones?)

### Required

- [x] I documented my function using [ROxygen style](https://jozef.io/r102-addin-roxytags/#:~:text=Inserting%20a%20skeleton%20%2D%20Do%20this,Shift%2BAlt%2BR%20).)

- [x] All parameters for the function are described in the documentation and the function has a decriptive title. 

Example:
```
#' Use GISTIC2.0 scores output to reproduce maftools::chromoplot with more flexibility
#'
#' @param scores output file scores.gistic from the run of GISTIC2.0
#' @param genes_to_label optional. Provide a data frame of genes to label (if mutated). The first 3 columns must contain chromosome, start, and end coordinates. Another required column must contain gene names and be named `gene`. (truncated for example)
#' @param cutoff optional. Used to determine which regions to color as aberrant. Must be float in the range [0-1]. (truncated for example)
```

- [ ] My function uses a library that isn't already a dependency of GAMBLR and I made the package aware of this dependency using the function documentation `import` statment. 

Example:
```
#' @return nothing
#' @export
#' @import tidyverse ggrepel
```

## Checklist for changes to existing code

- [ ] I added/removed arguments to a function and updated documentation for all changed/new arguments

- [ ] I tested the new code for compatability with existing functionality in the Master branch (please provide a reprex of how you tested the original functionality)

### CODE REVIEW

File: R/annotation.R
Comment (rdmorin) [inline R/annotation.R:275]: Is the maf altered in any way? Are there any new columns?

File: R/checks.R
Comment (rdmorin) [inline R/checks.R:47]: Does it return anything? I think maybe a boolean TRUE/FALSE depending on whether there is a remote connection?

File: R/checks.R
Comment (rdmorin) [inline R/checks.R:204]: would adding glue to this mean we don't have to put glue::glue everwhere?

File: R/checks.R
Comment (rdmorin) [inline R/checks.R:432]: "will notify the user and end the program if..."

File: R/database.R
Comment (rdmorin) [inline R/database.R:15]: I've seen you refer to a vector as a "list" many times. There is an important distinction between the two (at least in R) so please be sure you describe variables using the right class name. It should be "a vector of sample IDs" and that is probably almost always what you mean when the documentation says "list"

File: R/database.R
Comment (rdmorin) [inline R/database.R:50]: these_patient_ids A vector of sample_id that you want results for. 
The above argument should not be required if these_samples_metadata is provided. This is how every GAMBLR function should work. If the ids vector is missing, the function should get it from the metadata as long as the user provided it. 

these_samples_metadata Metadata table. It is expected the user has subset this to contain the rows corresponding to the patients of interest.

File: R/database.R
Comment (rdmorin) [inline R/database.R:139]: As in my last comment, please confirm the function works with either these_sample_ids or these_samples_metadata such that either of them can be provided to ensure the function runs only on the IDs the user wants. As long as this is how it works, please update the documentation so this is clear.

**Review by rdmorin — CHANGES_REQUESTED** [review]

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE]: This mistake was most likely due to an auto-merge, where `prettyForestPlot` had two import fields. I have removed the redundant @import section from this function (the one without broom specified) and re-generated the package docs. GAMBLR now imports broom, as it should. Good catch!

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R:7]: Got it! This has ben updated.

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R:7]: It's because I added punctuation at the end of every parameter description.

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R]: Yes, that makes sense. The example has been updated in this regard.

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R:136]: `separate` is from `tidyr`.

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R:127]: I would like you (or anyone else) to explain this function. Currently, it does not do anything, thus the return has been declared as "Nothing". Alternatively, I could remove the export for this function so that it doesn't populate the NAMESPACE.

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R]: There are no new columns, only a subset of the rows (mutations) identified as putative driver mutations. I have updated the return to state this clearly and expanded the example for the same function.

File: R/checks.R
Comment (mattssca) [reply R/checks.R]: Yes, I have updated the return description for this function (it actually returns the Boolean for `check_host`).

File: R/checks.R
Comment (mattssca) [reply R/checks.R:204]: Yes, based on my testing. But I have not applied these changes globally. 

There are basically two ways to tell a function what packages are needed. First, tell the function under the import section (in the documentation). Or, directly specify the package when a function from an external package is being called, followed by double colons (i.e `package::function`). Both options are valid and should achieve the same thing, at least from the perspective of the main function. Not sure how it would affect the overall documentation. I am thinking that the second option would mean that the external package would not show up under imports in the NAMESPACE.

File: R/checks.R
Comment (mattssca) [reply R/checks.R]: Nice, I have updated the docs for this function.

File: R/database.R
Comment (mattssca) [reply R/database.R]: Right, good point. I have gone through the complete package and updated this where needed.

File: R/database.R
Comment (mattssca) [reply R/database.R]: Yes, 100%. I have updated the parameter descriptions, function details and examples to make this clear. The function works as advertised with the two parameters discussed.

File: R/database.R
Comment (mattssca) [reply R/database.R]: I can confirm it works as advertised. I have also updated the documentation to make this clear as well as added an additional example demonstrating how to use this function with an already subset metadata table.

File: NAMESPACE
Comment (Kdreval) [inline NAMESPACE:142]: which package imports stats? We should drop this because it introduces conflicts with function names and gives lots of warnings on package installation/importing

File: NAMESPACE
Comment (Kdreval) [inline NAMESPACE:141]: I might misremember this - but I think there was some issue with installing the `ssh` package on GSC. So we opted to not import it and leave it to remote users to take care of. Maybe I am not correct or this has been resolved - but I would keep this out of the dependency list

File: NAMESPACE
Comment (Kdreval) [inline NAMESPACE:139]: Which function imports reshape2? I just dropped it in the latest mster update but seems like it is used somewhere else, too?

File: NAMESPACE
Comment (Kdreval) [inline NAMESPACE:82]: Not sure what happened with these 4 functions that were discontinued - but trust it was discussed somewhere. Just want to double-check with you it is intentional

File: NAMESPACE
Comment (Kdreval) [inline NAMESPACE:122]: which function was using forcats? Maybe we can rework it and avoid introducing this as dependency?

File: NAMESPACE
Comment (Kdreval) [inline NAMESPACE:132]: What is using gridExtra? Maybe we should rework that function to avoid another dependency? It is also probably already a sub-dependency of one of the packages like Complexeatmap or circlize

File: R/annotation.R
Comment (Kdreval) [reply R/annotation.R:7]: Did you do it by hand or there are some scripts/functions? I think it might not be sustainable in a long-term future unless we find a way to consistently follow this. It might be that it is now consistently introduced everywhere but the new functions will still be missing these so will then end up with a need another review/PR and it will keep going on like this. Just thinking if we can automate the things like punctuation.

File: R/annotation.R
Comment (Kdreval) [reply R/annotation.R:127]: You can compare the dimensions of the data before and after this function is run. Note: to see what it is doing you need to get the maf that has not been deblacklisted before, otherwise your input will be identical to the output since this functionality is used in [lcr-modules](https://github.com/LCR-BCCRC/lcr-modules/blob/78c3ea60e0b8cc2c27595b6dc7041aff2b71d9f9/modules/vcf2maf/1.3/vcf2maf.smk#L221-L244) to deblacklist mafs at the first place.

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:142]: Agreed. Assuming you're meaning "what function* imports stats", these are the functions that are currently importing stats;

```
FtestCNV
prettyForestPlot
prettyOncoplot
```

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:141]: Right, thanks for pointing this out. I will make the changes necessary

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:139]: Yes, it's imported by `prettyForestPlot`.

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:82]: Yes, these functions are still in GAMBLR, I just removed the @export for them, since they are all either in draft mode or/and are not yet doing anything. I thought this was the best approach since it does not make sense for a user to have access to functions that are not doing anything. Especially since this package is now open to the public. I also talked this over with Ryan when he was at BC Cancer a week ago, and he agreed. Thanks for checking!

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:122]: Currently, the only package that is importing `forcats` is `prettyForestPlot`.

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:132]: Currently, there's only one package that imports gridExtra and it is `comp_report`. It's the wrapper function for all the sample-level fancy plots. It uses `gridExtra` to arrange all plots and output a pdf. I can look into if the same output can be achieved without the dependency on this package.

File: NAMESPACE
Comment (Kdreval) [reply NAMESPACE:139]: I think your branch is not up to date with master. This was re-worked to remove the reshape2 dependency. The same also applies to stats package - in the recent master it is no longer imported

File: NAMESPACE
Comment (Kdreval) [reply NAMESPACE:82]: Thanks for confirming!

File: R/utilities.R
Comment (Kdreval) [inline R/utilities.R:2900]: The utils is a default R package always available when R starts. Do we need to explicitly import it here? Also, couldn't see where it is used below

File: R/viz.R
Comment (Kdreval) [inline R/viz.R:373]: This seems to be a duplicate of what is below. Wonder if this is intentional?

File: R/viz.R
Comment (Kdreval) [inline R/viz.R:403]: Same here for utils, wonder if it should be explicitly mentioned or  we can avoid it?

File: R/viz.R
Comment (Kdreval) [inline R/viz.R:5171]: Seems like this is just space additions?

**Review by Kdreval — COMMENTED** [review]
Thanks, Adam, so much work and important updates! I had some other questions/suggestions

File: NAMESPACE
Comment (Kdreval) [reply NAMESPACE:139]: The `stats` package is a default R package and is available when R starts. Therefore, to use `p.adjust`, we do not need to refer to the package specifically. Including the `stats`  in NAMESPACE explicitly only created name conflicts and many warnings generated once GAMBLR is loaded.

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:141]: For this, the code is still using functions from the `stats` package (from my testing).

- FtestCNV calls `fisher.test` and `adjust.p`
- prettyForestPlot calls `adjust.p`
- prettyOncoplot calls `setNames`

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:132]: Ok nice! Didn't know that `ggpubr` could achieve the same thing. I'll leave it as is for now.

File: R/checks.R
Comment (mattssca) [reply R/checks.R:120]: For this, I just added the newly added `check_config_value` to any **existing** line that gets paths from the config. So, if this indeed is something funny going on here, it would be outside the scope of what's introduced in this PR. But should be addressed nonetheless.

File: R/checks.R
Comment (mattssca) [reply R/checks.R:129]: Same here :)

File: R/database.R
Comment (mattssca) [reply R/database.R:642]: Yes, this solved a previous warning message that prevented the vignettes to knitted.

File: R/database.R
Comment (mattssca) [reply R/database.R:1242]: Yes, I thought this is something we wanted, to move away from the database. If not, I can easily add this back in 😄

File: R/database.R
Comment (mattssca) [reply R/database.R:1292]: The updates to this function are a combination of improving readability, as well s adding support for specifying multiple sample IDs as a vector. This is related to the issues you brought up in [this](https://morinlabsfu.slack.com/archives/C0224H120CU/p1674625364603939) Slack post. Hope it makes sense.

File: R/database.R
Comment (mattssca) [reply R/database.R:1823]: This is a good point, for now, I'll address this in an update. Thanks!

File: R/utilities.R
Comment (mattssca) [reply R/utilities.R:59]: Yeah, I think that's a good idea. I'll make the necessary updates. Thanks!

File: NAMESPACE
Comment (rdmorin) [reply NAMESPACE:139]: You could always update the code to say `stats::p.adjust()`	when it's called or add stats to the @imports line, right?

File: R/utilities.R
Comment (mattssca) [reply R/utilities.R:2900]: I'll go over all the code that currently imports `utils` based on the information you provided. In most cases, the functions called from this package are `head`, `tail`, `read_table` and `write_table`.

File: R/viz.R
Comment (mattssca) [reply R/viz.R:373]: Good catch, this is not intended.

**Comment by mattssca** [comment]
Thanks for your comments @Kdreval, I'll work on an update addressing the issues you raised. Stay tuned for a re-review!

File: NAMESPACE
Comment (mattssca) [reply NAMESPACE:139]: Ok, thanks for clarifying! Makes sense.

Comment (mattssca) [reply NAMESPACE:139]: FYI, after educating myself on what packages are default R packages (i.e loaded every time you start R), I found two additional packages that can be removed for the same reasons as `utils` and `stats`. These packages are `parallel` and `grid`.  I have removed them from being imported by any functions (i/e preventing them to populate NAMESPACE).

Comment (Kdreval) [reply NAMESPACE:139]: Thanks for checking!

File: R/annotation.R
Comment (Kdreval) [reply R/annotation.R]: I was wondering about this one here. It seems that the way example is set up it already loads the maf file that was generated as an output of this function. If you try this on the grande maf that is already bundled, does it annotates any variants to be deblacklisted? If so, we can use that to re-work the example here

File: R/database.R
Comment (Kdreval) [inline R/database.R:1359]: See this now

File: R/database.R
Comment (Kdreval) [reply R/database.R:1293]: I missed here that the other filtering steps were taken out of the if statement. Can you please confirm that the output after this modification is identical to what is currently returned in a version that is on master?

File: R/checks.R
Comment (rdmorin) [reply R/checks.R:116]: Wow. What does this line actually accomplish? I'm surprised R doesn't complain about this syntax.

Comment (mattssca) [reply R/checks.R:116]: This line creates two variables (local_base and base_path) pointing to the same path ("/projects/nhl_meta_analysis_scratch/gambl/results_local/"). 

Yeah, it's kinda strange that R accepts this. 

From the looks of it, this double assignment was added 7 months ago by Ryan. 

I will drop the `base_path` part since it, from the looks of it, doesn't do anything in the function. 

Sounds good?

File: R/annotation.R
Comment (rdmorin) [reply R/annotation.R:7]: My understanding is that Adam is now using a spellchecking add-on for VS Code. Maybe we should all try to use that if we're writing documentation?

Comment (Kdreval) [reply R/annotation.R:7]: Yes! Let's try to adopt it! Adam can you advertise this extension on Slack so we all will use the same add on (and any settings that need to be modified from default)? Thanks!

File: R/checks.R
Comment (rdmorin) [reply R/checks.R:116]: Yes it must be a pasting error. If you see something like this with weird syntax again please consider it a likely bug and fix it if you can.

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R:7]: Yes, It's working out great. The extension I am using is called Spell Right. I'll advertise it on Slack!

File: R/annotation.R
Comment (mattssca) [reply R/annotation.R]: This example has been updated to use the bundled grande MAF data (has not been deblacklsited).

**Review by Kdreval — APPROVED** [review]

**Review by rdmorin — APPROVED** [review]

**Pull request merged**
